p2p/host/basic/basic_host_notifiee.go

```diff
@@ -0,0 +1,54 @@
+package basichost
+
+import (
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/network"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// basicHostNotifiee listens to Network Notifiee events & emits them on the host's event bus
+type basicHostNotifiee BasicHost
+
+func (b *basicHostNotifiee) basicHost() *BasicHost {
+	return (*BasicHost)(b)
+}
+
+func (b *basicHostNotifiee) Connected(n network.Network, v network.Conn) {
+	p := v.RemotePeer()
+
+	// use the last byte of the peer as the key for the striped lock
+	indexForLk := len(p) - 1
+	lk := &b.stripedConnNotifLocks[p[indexForLk]]
+	lk.Lock()
+	defer lk.Unlock()
+
+	if n.Connectedness(p) == network.Connected {
```

Comment:
Trying to write some documentation for this... I realized this isn't quite correct. We want:

![good](https://user-images.githubusercontent.com/310393/76118908-21d2ea00-5fa3-11ea-85cd-62afbc24be71.png)


But this code will give us:

![bad](https://user-images.githubusercontent.com/310393/76118921-27303480-5fa3-11ea-8511-c152909caf00.png)

---

We used code like this elsewhere because transitioning within the current state wasn't actually an issue. However, if we're going to make this a public interface, we shouldn't do this.

Your original code tried to avoid this by checking "are we transitioning from/to 0" but the asynchronous nature of these events meant that didn't work either. Unfortunately... I think we need to remember the last event emitted for the peer to fix this.